pass_commander/main.py

```diff
@@ -19,149 +19,136 @@
-edl_packet = confget(config, ["Main", "edl"])
-local_only = False
-no_tx = False
-no_rot = False
-tx_gain = int(confget(config, ["Main", "txgain"]))
-if len(edl_packet) <= 10:
-    print('Not going to TX because no EDL bytes have been defined')
-    no_tx = True
-
-# XXX These should be set by command line arguments
-# local_only=True # XXX Test mode with no connections
-# no_tx=True  # XXX
-# no_rot=True  # XXX
-
-tle_cache = {}
-tle_cache_file = os.path.expanduser("~/.config/OreSat/tle_cache.json")
-if os.path.isfile(tle_cache_file):
-    with open(tle_cache_file, "r") as jsonfile:
-        tle_cache = json.load(jsonfile)
-
-log.basicConfig()
-log.getLogger("apscheduler").setLevel(log.ERROR)
+import ephem
+import pydbus
+from apscheduler.schedulers.background import BackgroundScheduler
 
+from .Navigator import Navigator
+from .Radio import Radio
+from .Rotator import Rotator
+from .Station import Station
+from .Tracker import Tracker
+
+
+@dataclass
+class Config:
+    # Main
+    owmid: str
+    edl: str
+    txgain: int
+
+    # Hosts
+    radio: str
+    station: str
+    rotator: str
+
+    # Observer
+    lat: float
+    lon: float
+    alt: int
+    name: str
+
+    az_cal: int = 0
+    el_cal: int = 0
+
+    # Satellite
+    sat_id: str = "OreSat0"
+    tle_cache: dict[Union[list[str], bool]] = field(default_factory=dict)
+
+
+def load_config_file(path):
+    config_file = os.path.expanduser(path)
+    config = configparser.ConfigParser()
+    if not len(config.read(config_file)):
+        print("Config file seems to be missing. Initializing.")
+        if not os.path.exists(os.path.dirname(config_file)):
+            os.makedirs(os.path.dirname(config_file))
+        with open(config_file, "w") as f:
+            f.write(dedent("""\
+                # Be sure to replace all <hint text> including the angle brackets!
+                [Main]
+                owmid = <open weather map API key>
+                edl = <EDL command to send, hex formatted with no 0x prefix>
+                txgain = 47
+
+                [Hosts]
+                radio = 127.0.0.2
+                station = 127.0.0.1
+                rotator = 127.0.0.1
+
+                [Observer]
+                lat = <latitude in decimal notation>
+                lon = <longitude in decimal notation>
+                alt = <altitude in meters>
+                name = <station name or callsign>
+                """))
+        print(f"Please edit {config_file} before running again.")
+        sys.exit(1)
+
+    if '<' in [v[0] for s in config.keys() for v in config[s].values()]:
+        print(f"Please edit {config_file} and replace everything in <angle brackets>")
+        sys.exit(1)
+
+    # This is just shoehorned in here and ugly. Please fix!
+    def confget(conf, tree):
+        try:
+            return reduce(operator.getitem, tree, conf)
+        except KeyError:
+            print(f"Configuration element missing: {tree[0]}")
+            sys.exit(2)
```

Comment:
This seems really squirreley. As far as I can understand this is doing the equivalent to `getattr(conf, tree)` but without the default field. Is there not a reason we could handle this gracefully and provide a default that would have been used in the template above instead of crashing?

```suggestion
    def confget(conf, tree, default):
        return getattr(conf, tree, default)
```

Reply:
Close, given the args it's called with it is doing `conf.__getitem__(tree[0]).__getitem__(tree[1])`. Cute, but also yeah weird and not strictly correct. If I was to replace it I'd just do the access directly, e.g. `config['Main']['owmid']` and catch the `KeyError` on missing entries.

I left it as-is to keep behavior as close to how it was before. I would like to do another PR to overhaul config file handling eventually and it would be appropriate there.